README.md
# requestmq

## Como usar

Adicionar no `package.json` a dependência `requestmq`:

    "requestmq": "git+https://github.com/gustavopaes/requestmq.git"

No seu projeto, utilizar como um módulo qualquer:

```
const requestmq = require('requestmq');

requestmq({
    host: process.env.RABBIT_HOST,
    queue: process.env.RABBIT_QUEUE,
    confirmChannel: true,
    prefetch: process.env.RABBIT_PREFETCH || 5
})
.consume(item => console.log('Item consumido: ', item))
.sendToQueue({...json com dados do request...});
```

## Formato JSON

O JSON enviado para o RabbitMQ deve seguir um padrão, para que o `requestmq` consiga interpretar
e fazer corretamente o request.

Segue exemplo de JSON que indica um request tipo POST, com dados a serem enviados no body do request:

    {
      request: {
        protocol: 'https:',
        hostname: 'graph.facebook.com',
        path: '/v2.6/me/messages?access_token=XXX',
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        timeout: 5000
      },
      post: {
        foo: "dado que será enviado",
        bar: {
            zza: "via header do request"
        }
      }
    }

## Variáveis de ambiente

### Proxy

    HTTPS_PROXY
    url:port para proxy


### RabbitMQ

    RABBIT_HOST
    Host para o RabbitMQ -- com usuário, senha e porta. Não adicionar protocolo.
    Exemplo: admin:admin@rabbitmqhost.intranet
    
    RABBIT_QUEUE
    Nome da fila
    
    RABBIT_PREFETCH
    Prefetch dos workers (default 5)


CHANGELOG
[1.0.11] 06/06/2017
- Verifica se foi passado "url" do proxy

[1.0.10] 13/03/2017
- Utiliza variável de ambiente HTTPS_PROXY para definição de URL e Porta de proxy

[1.0.9] 23/02/2017
- Faz log dos headers de debug do facebook
- Reenvia para a fila Rabbit (nack) requests com erro 502 (bad gateway)

[1.0.8] 21/02/2017
- Melhorias nos logs

[1.0.7] 20/02/2017
- Faz o log do erro do consumer
- Registra ao menos um worker (cpu) de consumo

[1.0.6] 17/02/2017
- Passando objeto do request original para o master

[1.0.5] 17/02/2017
- Repassa o retorno de um request para o master, que irá repassar para
  uma função callback (se definida pelo script que chamou o consume())

[1.0.4] 16/02/2017
- Mata os workers quando o master morre
- Cria esse CHANGELOG

[1.0.1] 15/02/2017
- Implementa configuração de proxy no request

[1.0.0] 13/02/2017
- Primeira versão

requestmq.js
'use strict';

const debug = require('debug')('[requestmq]');
const error = require('debug')('[requestmq] Error');
const CPUS = (require('os').cpus().length - 1) || 1;
const fork = require('child_process').fork;
const joinPath = require('path').join;
const consumer = require('./consumer');

const mqueue = require('amqplib');
const MQ_CONN = undefined;
const MQ_CHANNEL = undefined;

// total de requests paralelos para enviar itens para a fila MQ
const MAX_SEND_MQ = CPUS * 15;

// total de mensagens que um worker pode receber
const MAX_WORKER_MESSAGES = 10;

const uniqueId = () => process.hrtime().join('');

class RequestMQ {
  constructor(config) {
    config = config || {};

    this.conn = undefined;
    this.channel = undefined;
    this.consumeCallback = undefined;
    this.canConsume = false;
    this.isConsuming = false;

    // quantidade de requests paralelos que estão sendo feitos
    // enviando itens da fila local para a fila MQ
    this.publish_workers_working = 0;

    // workers de publicação (consumers)
    this.workers = [];

    // Nome da fila no MQ
    this.config = config;

    // Fila de itens que devem ser enviados para a fila MQ
    this.queue = [];

    // Indica se os itens da fila local estão sendo enviados
    // para a fila MQ.
    this.isSendingToQueue = false;

    // Registra evento para remover todos os workers quando a aplicação
    // for encerrada.
    process.on('uncaughtException', err => {
      error('uncaughtException', err);
      //this.stopConsume();
      //process.exit(1);
    });

    process.on('exit', () => {
      debug('exit');
      this.stopConsume();
      process.exit(0);
    });

    process.on('SIGINT', () => {
      debug('SIGINT');
      this.stopConsume();
      process.exit(0);
    });

    process.on('SIGTERM', () => {
      debug('SIGTERM');
      this.stopConsume();
      process.exit(0);
    });

    // Faz a conexão, cria o channel e efetua o assert da fila
    return this.connect();
  }

  connect() {
    return mqueue.connect(`amqp://${this.config.host}/`)
      .then(conn => {
        this.conn = conn;

        // registra eventos de erro
        conn.on('error', err => {
          this.conn = undefined;
          this.channel = undefined;

          error('Erro com a conexão rabbit');
          error(err);

          // zera o contador de requests paralelos para o rabbit
          // para que a fila começe novamente
          this.publish_workers_working = 0;
          this.sendToMQ();

          // tenta conectar novamente depois de alguns segundos
          setTimeout(this.connect.bind(this), 10 * 1000);
        });

        return Promise.resolve(conn);
      })
      .then(conn => {
        if(this.config.confirmChannel === true) {
          return this.conn.createConfirmChannel();
        } else {
          return this.conn.createChannel();
        }
      })
      .then(channel => {
        this.channel = channel;

        // define o prefetch
        channel.prefetch(parseInt(this.config.prefetch || 5));

        return channel.assertQueue(this.config.queue).then(ok => {
          return Promise.resolve(this);
        });
      })
      .catch(err => {
        error(`Erro ao conectar em amqp://${this.config.host}`);
        error(err);

        // tenta conectar novamente depois de alguns segundos
        setTimeout(this.connect.bind(this), 10 * 1000);

        return Promise.reject(this);
        //process.exit(1);
      });
  }

  /**
   * Registra um consumidor para a fila.
   * @param {Function} cb Função que será executada ao receber um item para consumo
   */
  consume(cb) {
    if(this.isConsuming === true) {
      return Promise.resolve(this);
    }

    this.consumeCallback = cb;
    this.canConsume = true;
    this.isConsuming = true;

    // cria os workers de consumo
    for(let i = 0; i < CPUS; i++) {
      let worker = {
        id: i,
        busy: false,
        fork: fork(joinPath(__dirname, './consumer.js')),
      };

      worker.fork.send(this.config);

      if(typeof cb == 'function') {
        worker.fork.on('message', cb);
      }

      this.workers.push(worker);
    }

    return Promise.resolve(this);
  }

  stopConsume() {
    this.isConsuming = false;

    if(this.workers.length) {
      // mata todos os workers de consumo, para a publicação
      // dos itens na fila ficarem trabalhando sozinhos
      this.workers = this.workers.filter(w => {
        w.fork.kill('SIGHUP');

        return false;
      });
    }
  }

  /**
   * Envia um sinal de acknowledge para o MQ informado que o item foi
   * consumido com sucesso e que ele pode ser removido da fila.
   * @param {Message} Objeto da mensagem recebida
   */
  ack(item) {
    this.channel.ack(item);
    return Promise.resolve(this);
  }

  /**
   * Adiciona à fila local um item que deve ser enviado para a fila MQ.
   * @param {Object} item
   */
  sendToQueue(item) {
    item.id = uniqueId();

    debug(`${item.id} adicionando item à fila local`);
    debug(`${item.id}`, JSON.stringify(item));
    this.queue.push(item);

    if(this.isSendingToQueue === false || this.sending_mq < MAX_SEND_MQ) {
      this.stopConsume();
      this.sendToMQ();
    }

    return Promise.resolve(this);
  }

  /**
   * Envia para a fila MQ os itens que estão na fila local.
   */
  sendToMQ() {
    if(this.queue.length === 0) {
      this.isSendingToQueue = false;
      this.sending_mq = 0;

      // inicia workers de consumo
      if(this.canConsume === true) {
        this.consume(this.consumeCallback);
      }

      return false;
    }

    // já chegou ao limite de publicações paralelas
    if(this.publish_workers_working >= MAX_SEND_MQ) {
      debug(`máximo de envios (${MAX_SEND_MQ}) paralelos atingido`);
      return false;
    }

    this.isSendintToQueue = true;
    this.publish_workers_working += 1;

    let item = this.queue.pop();
    let data = typeof item === 'string' ? item : JSON.stringify(item);

    debug(`${item.id}\tiniciando envio de item para o rabbit`);

    // se não há conexão com o rabbit, faz a publicação
    // diretamente para o serviço
    if(this.channel === undefined) {
      consumer
        .request(item)
        .then(this.delivered.bind(this))
        .catch(err => {
          this.sendToQueue(item);
          this.delivered();
        })

      return false;
    }

    this.channel.sendToQueue(this.config.queue, new Buffer(data), {}, (err, ok) => {
      if(err) {
        debug(`${item.id}\terro ao enviar item ao rabbitmq`, err);
        this.sendToQueue(item);
      }

      debug(`${item.id}\titem enviado para o rabbitmq com sucesso`);
      this.delivered();
    });
  }

  delivered() {
    this.publish_workers_working -= 1;
    this.sendToMQ();
  }
};

module.exports = (host, config) => new RequestMQ(host, config);


consumer.js
'use strict';

const debug = require('debug')('[requestmq] Worker ' + process.pid);
const error = require('debug')('[requestmq] Error');
const requestmq = require('./requestmq.js');
const https = require('https');
const http = require('http');
const parseUrl = require('url').parse;

const config = require('./package.json').config;
const httpsProxyAgent = require('https-proxy-agent');

let TIME_REQUESTS = 0;
let TOTAL_REQUESTS = 0;

// Configurações do proxy
let proxyConfig = process.env.HTTPS_PROXY || '';

const connect = config => {
  if(config.proxy && config.proxy.url) {
    // sobrescreve a configuração inicial do proxy pela config
    // do utilizador
    proxyConfig = `${config.proxy.url}:${config.proxy.port}`;
  }

  return requestmq(config);
};

const parseContent = content => {
  try {
    return JSON.parse(content.toString());
  } catch(err) {
    error('Erro ao executar JSON.parse em mensagem do Rabbit', err);
    return null;
  }
}

const consume = queue => {
  if(queue.channel) {
    return queue.channel.consume(queue.config.queue, message => {
      let data = parseContent(message.content);
      let method = (data.request.method || 'get').toLowerCase();

      request(data)
        .then(() => {
          debug(`${data.id}\trabbit ack`);
          queue.channel.ack(message);
        })
        .catch((err) => {
          debug(`${data.id}\trabbit nack`, err);
          queue.channel.nack(message);
        });
    });
  } else {
    return Promise.resolve(queue);
  }
};

const post = data => {
  return new Promise((resolve, reject) => {
    // define proxy para conexão post
    if(proxyConfig !== '') {
      data.request['agent'] = new httpsProxyAgent( proxyConfig ); 
    }

    let requestStartTime = new Date();

    let protocol = https;
    if(data.request.protocol == 'http:') {
      protocol = http;
    }

    var req = protocol.request(data.request, function(res) {
      res.setEncoding('utf8');
      let body = '';

      debug(`${data.id}\tstatus code: ${res.statusCode}.`);

      res.on('data', chunck => {
        body += chunck;
      });

      res.on('end', () => {
        debug(`${data.id}\tbody response: `, body);
        debug(`${data.id}\theader response: `, res.headers);

        switch(res.statusCode) {
          case 502:
            reject(body);
            break;

          default:
            resolve(body);
            break;
        }

        // cria média de tempo de request
        TIME_REQUESTS += (new Date() - requestStartTime);
        TOTAL_REQUESTS += 1;
      });
    });

    req.on('error', function(err) {
      debug(`${data.id}\terro no request:`, err);

      req.abort();
      reject(err.message);
    });

    if(data.post !== undefined) {
      debug(`${data.id}\tpost body`, JSON.stringify(data.post));
      req.write(JSON.stringify(data.post));
    }

    req.end();
  });
};

const request = data => {
  debug(`${data.id}\tFaz request ${data.request.method} para ${data.request.hostname}`);

  if(data.request.method == 'POST') {
    return post(data)
            .then(response => {
              process.send({ response: response, data: data });
              return Promise.resolve(response);
            })
            .catch(response => {
              return Promise.reject(response);
            });
  } else {
    //return get(data);
  }
};

process.on('message', m => {
  connect(m).then(consume);
});

setInterval(() => {
  if(TOTAL_REQUESTS) {
    debug('tempo médio para publicações: %d (requests: %d)', TIME_REQUESTS / TOTAL_REQUESTS, TOTAL_REQUESTS);
  }

  if(TOTAL_REQUESTS > Number.MAX_SAFE_INTEGER / 2) {
    // Zera o valores de estatísticas
    TIME_REQUESTS = 0;
    TOTAL_REQUESTS = 0;
  }
}, 60 * 1000);

module.exports = {
  request
};
